### PR TITLE
[charts] Move interaction state in store

### PR DESCRIPTION
**Pull request by alexfauquette**
Extracted from #15154 focusing only on the introduction of the store.

The PR can be read commit by commit.

The performances gain for the tooltip are not interesting in this PR, because I tried to keep it without breaking change. And the props of the axis tooltip are too generous. They provide the state for x and y axis. So as soon a the mouse move I need to update the state either for x or y. The performances of the tooltip will be done in the breaking change of  #15154

The axis highlight get some performances improvement by having a component for the x-axis and another for the y-axis. Such that they re-render only if the correct value is updated.

@JCQuintas I noticed your concern about the naming of `InteractionProvider`. The plan is to go one provider per PR, and move all the typing and the logic into the `pluggin` folder. At the end it will be named `ChartsDataProvider` or whatever you name the component in your PR to split the ChartContainer

### CODE REVIEW

File: packages/x-charts/src/internals/plugins/utils/selectors.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  Map<Parameters<typeof reselectCreateSelector>, any>
+>();
+
+export type ChartsRootSelector = (state: ChartsState) => ChartsState[keyof ChartsState];
+
+export type ChartsSelector<TState, TArgs, TResult> = (state: TState, args: TArgs) => TResult;
+
+/**
+ * Method wrapping reselect's createSelector to provide caching for tree view instances.
+ *
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/utils/selectors.ts:23]:
```suggestion
 * Method wrapping reselect's createSelector to provide caching for chart instances.
```

File: packages/x-charts/src/internals/plugins/models/index.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+   * Set to `true` when `VoronoiHandler` is active.
+   * Used to prevent collision with mouseEnter events.
+   */
+  useVoronoiInteraction?: boolean;
+};
+
+export type ChartsStateCacheKey = { id: number };
+
+export type ChartsState = {
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/plugins/models/index.ts:36]: I think singular `ChartState`/`ChartStore` would sound better here 🙃

File: packages/x-charts/src/internals/useStore.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import { ChartsContext } from '../context/InteractionProvider';
+import { ChartsStore } from './plugins/utils/ChartsStore';
+
+export function useStore(skipError?: boolean): ChartsStore {
+  const charts = React.useContext(ChartsContext);
+
+  if (skipError) {
+    // This line is only for `useAxisEvents` which is in the surface of the Gauge.
```
Comment (JCQuintas) [inline packages/x-charts/src/internals/useStore.ts:9]:
```suggestion
    // TODO: Remove once store is used by all charts.
    // This line is only for `useAxisEvents` which is in the surface of the Gauge.
```

**Review by JCQuintas — COMMENTED** [review]
Nice extraction, everything seems to work nicely. 

I've left some comments for possible improvements.

I would also argue that everything new created inside the `internals/` folder could either be in a root `store/` folder or inside `internals/store/`.

Since the `internals` is quite bloated already without specific access patterns 😓

File: packages/x-charts/src/hooks/useAxisEvents.ts [hunk 4]
```diff
@@ -121,7 +129,24 @@ export const useAxisEvents = (disableAxisListener: boolean) => {
...
+            prev.interaction.axis.x?.value !== newStateX?.value
+              ? { x: newStateX }
+              : {}),
+            ...(prev.interaction.axis.y?.index !== newStateY?.index ||
+            prev.interaction.axis.y?.value !== newStateY?.value
+              ? { y: newStateY }
+              : {}),
+          },
+        },
+      }));
```
[the remark this answers is not in the document]
Comment (alexfauquette) [reply packages/x-charts/src/hooks/useAxisEvents.ts:149]: Does not work. I suspect that mutation `prev.interaction.axis.x = newStateX;` is such that it mutates also the previouse state, and then cause the pipeline to do nto realise the `interaction.axis.x` got updated.


https://redux.js.org/tutorials/fundamentals/part-8-modern-redux#immutable-updates-with-immer

**Comment by alexfauquette** [comment]
About renaming files, I would prefer to do that in a dedicated PR after merging this one and #15154

**Review by JCQuintas — APPROVED** [review]

**Pull request merged**